cmd/agent/api/server.go

```diff
@@ -45,16 +45,24 @@ func grpcHandlerFunc(grpcServer *grpc.Server, otherHandler http.Handler) http.Ha
 		if r.ProtoMajor == 2 && strings.Contains(r.Header.Get("Content-Type"), "application/grpc") {
 			grpcServer.ServeHTTP(w, r)
 		} else {
-			deadline := time.Now().Add(config.Datadog.GetDuration("server_timeout") * time.Second)
-
-			conn := agent.GetConnection(r)
-			_ = conn.SetWriteDeadline(deadline)
-
 			otherHandler.ServeHTTP(w, r)
 		}
 	})
 }
 
+// timeoutHandler limits requests to the enclosed handler to the value
+// configured in `server_timeout`.
+func timeoutHandlerFunc(otherHandler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		deadline := time.Now().Add(config.Datadog.GetDuration("server_timeout") * time.Second)
+
+		conn := agent.GetConnection(r)
+		_ = conn.SetWriteDeadline(deadline)
+
+		otherHandler.ServeHTTP(w, r)
```

Comment:
Wouldn't `otherHandler.ServeHTTP` get called twice now? Once here and once in `grpcHandlerFunc` on the same instance?

Reply:
The `otherHandler` in `grpcHandlerFunc` _is_ this handler func.  So, they're nested: `net/http` calls the handler from `grpcHandlerFunc`, which calls the handler from `timeoutHandlerFunc`, which calls the handler from the Mux.

Reply:
Makes sense. Wasn't immediately clear to me how that behaved. 

Just my 2 cents, but I actually found the logic here easier to follow when it was embedded in the conditional as it was more clear exactly which case it applied to and when it gets called. But if wrapping handlers is the more idiomatic thing to do, feel free to disregard this feedback.

Reply:
I'm not an expert on idiomatic go, but the blog posts I've read about the design of the Handler interface suggest that this kind of layers-of-middleware structure is the intention.  But, I can add a comment to where these functions are called to help explain the intent.